pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -592,19 +659,42 @@ func (r *ReconcileBareMetalHost) manageHostPower(prov provisioner.Provisioner, i
 		return actionContinue{provResult.RequeueAfter}
 	}
 
+	pendingRebootSince := info.host.Status.Provisioning.PendingRebootSince
+	lastPoweredOn := info.host.Status.Provisioning.LastPoweredOn
+
+	desiredPowerOnState := info.host.Spec.Online
+	isInRebootProcess := lastPoweredOn.Before(pendingRebootSince) || (pendingRebootSince != nil && lastPoweredOn == nil)
+
+	if isInRebootProcess {
+		suffixlessAnnotationExists, shouldHoldPowerOff := getRebootAnnotations(info.host)
+
+		if info.host.Status.PoweredOn || shouldHoldPowerOff {
+			desiredPowerOnState = false
+		} else {
+			if suffixlessAnnotationExists {
+				delete(info.host.Annotations, rebootAnnotationPrefix)
+
+				if err = r.client.Update(context.TODO(), info.host); err != nil {
```

Comment:
If we save the resource here and then need to save some other value again later, we'll end up with a conflict error. We usually wait to save the host in Reconcile() based on the type of return value from this method to avoid that race and to allow us to make multiple changes to the resource. This is updating the resource, not it's status, so the existing call to save the resource won't work.

I think to avoid the race, the logic to manage the annotations needs to all happen in this method, before the power is toggled (see related comment about the changes in Reconcile() above).

If the annotation is present and pendingRebootSince is not set, set pendingRebootSince and return actionContinue immediately so Reconcile() stores the status and then loops back to this point to change the power status.

If the annotation is present the lastPoweredOn value is after pendingRebootSince, and the power is actually on, remove the annotation, call Update(), and return actionContinue immediately, without making any other changes. 

Otherwise, figure out what the power state should be and try to enforce it.

Restructuring like that means that we will enter this method a few different times as part of handling each step of the reboot, but a failed attempt to power back on as part of a reboot will keep trying and the annotation will only be removed when the power on call is successful.

Reply:
Good point! I replaced the `Update()` with `return actionContinue{}`.
Meanwhile I keep `checkUpdatedRebootRequest()` in it's place until we decide otherwise in the previous comment

Let me know if this covers everything or something is still missing

Reply:
Just returning isn't going to work, because it won't save the update to the annotation. That's why I tried to explain how to manage the annotations and avoid races from updating the metadata and status in the same pass through the reconcile loop.